src/main/java/fi/aalto/cs/apluscourses/intellij/model/APlusProject.java

```diff
@@ -40,43 +48,38 @@ public LibraryTable getLibraryTable() {
     return LibraryTablesRegistrar.getInstance().getLibraryTable(project);
   }
 
-  @NotNull
-  public Path getLibraryPath(String name) {
-    return  Paths.get("lib", name);
-  }
-
-  @NotNull
-  public Path getModulePath(String name) {
-    return Paths.get(name);
-  }
-
-  public int resolveLibraryState(String name) {
-    return resolveComponentState(getLibraryTable().getLibraryByName(name), getLibraryPath(name));
-  }
-
-  public int resolveModuleState(String name) {
-    return resolveComponentState(getModuleManager().findModuleByName(name), getModulePath(name));
-  }
-
-  private int resolveComponentState(@Nullable Object componentObj, @NotNull Path dirPath) {
+  /**
+   * Returns a state in which the given component should be.
+   * @param component An IntelliJ specific component.
+   * @param <T> Type of the component.
+   * @return Component state.
+   */
+  public <T extends Component & IntelliJComponent<?>> int resolveComponentState(
```

Comment:
so...

1. `<?>` is https://docs.oracle.com/javase/tutorial/extra/generics/wildcards.html
2. `<T extends Component & ... > is https://docs.oracle.com/javase/tutorial/java/generics/boundedTypeParams.html

Reply:
I wanted the parameter `component` to be both `Component` and `IntelliJComponent`. If `IntelliJComponent` was a subtype of `Component` (as it semantically should be) it would be enough to have `public int resolveComponentState(IntelliJComponent<?> component)`.

The problem is that `Component` is an abstract class and `IntelliJComponent` is an interface, and Java doesn't allow interface to inherit an abstract class (that is, restrict interface to be implemented only by subclasses of `Component`, which actually would kinda make sense).

I guess the "most correct" way here would be making `Component` an interface too (and having a separate `ComponentBase` base class for implementations). Then `IntelliJComponent` could just extend `Component`.

Reply:
But yeah this solution (which abuses type parameters) is awkward, especially if there will be more methods like this in the future.